testr/runner.py

```diff
@@ -121,3 +126,31 @@ def chdir(dirname=None):
         raise TestError('got {} failure(s)'.format(n_fail))
 
     return n_fail
+
+
+def get_full_version(calling_frame_globals, calling_frame_filename):
+    """
+    Return a full version which includes git info if the module was imported
+    from the git repo source directory.
+    """
+    release_version = calling_frame_globals.get('__version__', 'unknown')
+
+    try:
+        from subprocess import Popen, PIPE
+        filedir = os.path.dirname(os.path.abspath(calling_frame_filename))
+
+        p = Popen(['git', 'rev-list', 'HEAD'],
+                  cwd=filedir,
+                  stdout=PIPE, stderr=PIPE, stdin=PIPE)
+        stdout, stderr = p.communicate()
+        stdout = stdout.decode('ascii')
+
+        if p.returncode == 0:
+            revs = stdout.split('\n')
+            out = release_version + '-r{}-{}'.format(len(revs), revs[0][:7])
```

Comment:
Not following you on a possible advantage.

Reply:
Well, `git describe --always` returns

last_tag n_commits_since_tag current_sha

in one step/line.  And doesn't fetch all the versions only to use the last one (if I'm understanding your Popen).
But last_tag might not match release_version if we're lazy.

Reply:
Yeah, I think it's pretty important to use the version that the package believes, not what git believes.  Also I want n_commits_since_start, not since last tag.  Actually trying out `git describe --always` on astropy master gives a curious result:
```
(astropy) neptune$ git describe --always
v1.0rc1-9428-gc04ced3
```
The actually version and number of commits since start is `3.1.dev-r21186`.  I guess astropy branches before tagging or something.

Reply:
Ah, then this mini discussion bore fruit anyway, as that's good to know.